File: src/algorithm/leetcode/Problem79.java
package algorithm.leetcode;

public class Problem79 {

    public static boolean existSearch(char[][] board, int curX, int curY,
                                      boolean[][] visited, StringBuilder curSb,
                                      int curIdx, String target) {
        if (curSb.length() == target.length())
            return true;
        if (curX >= board[0].length || curY >= board.length || curX < 0 || curY < 0) {
            return false;
        }
        if (visited[curY][curX] == true)
            return false;
        if (board[curY][curX] == target.charAt(curIdx)) {
            StringBuilder newSb = new StringBuilder(curSb);
            newSb.append(board[curY][curX]);
            visited[curY][curX] = true;
            boolean rst =  existSearch(board, curX + 1, curY, visited, newSb, curIdx + 1, target) ||
                    existSearch(board, curX - 1, curY, visited, newSb, curIdx + 1, target) ||
                    existSearch(board, curX, curY + 1, visited, newSb, curIdx + 1, target) ||
                    existSearch(board, curX, curY - 1, visited, newSb, curIdx + 1, target);
            visited[curY][curX] = false;
            return rst;
        } else {
            return false;
        }

    }

    public static boolean exist(char[][] board, String word) {
        if (board.length == 0)
            return false;
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                boolean[][] visited = new boolean[board.length][board[0].length];
                StringBuilder sb = new StringBuilder();
                if (existSearch(board, j, i, visited, sb, 0, word))
                    return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        char[][] board = new char[][]{
                {'A', 'B', 'C', 'E'},
                {'S', 'F', 'E', 'S'},
                {'A', 'D', 'E', 'E'}
        };

        System.out.println(exist(board, "ABCESEEEFS"));
//        char[][] board = new char[][]{
//                {'a', 'a'}
//        };
//
//        System.out.println(exist(board, "aaa"));
    }
}

File: src/algorithm_cpp/problem79.cpp
#include<vector>
#include<string>
#include<iostream>

using namespace std;

class Solution {
public:
    bool search(const vector<vector<char>>& board, vector<vector<bool>>& used, const string& word, int i, int j, int curidx) {
	if (curidx >= word.size())
	    return true;
	if (used[i][j] == true)
	    return false;
	if (board[i][j] == word.at(curidx)) {
	    int a = i - 1 < 0 ? 0 : i - 1;
	    int b = j - 1 < 0 ? 0 : j - 1;
	    int c = i + 1 > board.size() - 1 ? board.size() - 1 : i + 1;
	    int d = j + 1 > board[0].size() - 1 ? board[0].size() - 1 : j + 1;
	    used[i][j] = true;
	    bool rt = search(board, used, word, a, j, curidx + 1) || search(board, used, word, i, b, curidx + 1) || search(board, used, word, c, j, curidx + 1) || search(board, used, word, i, d, curidx + 1);
	    used[i][j] = false;
	    return rt;
	} else {
	    return false;
	}
    }

    bool exist(vector<vector<char>>& board, string word) {
	vector<vector<bool>> used;
	used.resize(board.size());
	for (vector<bool>& u : used)
	    u.resize(board[0].size());
	for (int i = 0; i < board.size(); i++) {
	    for (int j = 0; j < board[0].size(); j++) {
		if(search(board, used, word, i, j, 0) == true)
		    return true;
	    }
	}
	return false;
    }
};
			
	    
int main() {
    Solution s;
    vector<vector<char>> board {
      {'A','B','C','E'},
      {'S','F','C','S'},
      {'A','D','E','E'}
    };
    cout << s.exist(board, "ABCCED");
}
